core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -203,17 +203,44 @@ private[fs2] object FreeC {
     final case class View[F[_], X, R](step: F[X], next: Result[X] => FreeC[F, R])
         extends ViewL[F, R]
 
+    final case class >=>[F[_], X, Y, Z](fst: Result[X] => FreeC[F, Y],
+                                        snd: Result[Y] => FreeC[F, Z])
+        extends (Result[X] => FreeC[F, Z]) {
+      def apply(r: Result[X]): FreeC[F, Z] = Bind[F, Y, Z](fst(r), snd)
+    }
+
     private[fs2] def apply[F[_], R](free: FreeC[F, R]): ViewL[F, R] = mk(free)
 
     @tailrec
     private def mk[F[_], R](free: FreeC[F, R]): ViewL[F, R] =
       free match {
         case Eval(fx) => View(fx, pureContinuation[F, R])
         case b: FreeC.Bind[F, y, R] =>
-          b.fx match {
-            case Result(r)  => mk(b.f(r))
-            case Eval(fr)   => ViewL.View(fr, b.f)
-            case Bind(w, g) => mk(Bind(w, (e: Result[Any]) => Bind(g(e), b.f)))
+          @tailrec
+          def mk2[Y, Z](free: FreeC[F, Y],
+                        cont: Result[Y] => FreeC[F, Z]): Either[FreeC[F, Z], ViewL[F, Z]] =
+            free match {
+              case b: FreeC.Bind[F, x, Y] => mk2[x, Z](b.fx, >=>(b.f, cont))
+              case Eval(fx)               => Right(View(fx, cont))
+              case r @ Result.Pure(_) =>
```

Comment:
I did try extracting the `Result` cases in a single function, but then the `@tailrec` annotation would not work.